### PR TITLE
Make flow graph for delta channel

### PR DESCRIPTION
**Pull request by godchen0212**
Make flow graph for delta channel
Issue #11014 
/kind enhancement

Signed-off-by: godchen <qingxiang.chen@zilliz.com>

### CODE REVIEW

File: internal/querynode/flow_graph_delete_node.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ func (dNode *deleteNode) Operate(in []flowgraph.Msg) []flowgraph.Msg {
 	// 1. filter segment by bloom filter
 	for _, delMsg := range dMsg.deleteMessages {
 		if dNode.replica.getSegmentNum() != 0 {
+			log.Debug("delete in historical replica")
```
Comment (bigsheeper) [inline internal/querynode/flow_graph_delete_node.go:54]: Add more log info is recommended, such as `collectionID`,  `collectionName` and `timestamp`.

File: internal/querynode/flow_graph_insert_node.go [hunk 2]
```diff
@@ -149,16 +148,17 @@ func (iNode *insertNode) Operate(in []flowgraph.Msg) []flowgraph.Msg {
 	// 1. filter segment by bloom filter
 	for _, delMsg := range iMsg.deleteMessages {
 		if iNode.streamingReplica.getSegmentNum() != 0 {
+			log.Debug("delete in streaming replica")
```
Comment (bigsheeper) [inline internal/querynode/flow_graph_insert_node.go:151]: add more log info

File: internal/querynode/flow_graph_query_node.go [hunk 3]
```diff
@@ -210,6 +209,10 @@ func (q *queryNodeFlowGraph) seekQueryNodeFlowGraph(position *internalpb.MsgPosi
 	)
 	return err
 }
+func (q *queryNodeFlowGraph) startConsumer() {
+	log.Debug("start consumer", zap.Any("channel", q.channel), zap.Any("collectionID", q.collectionID))
+	q.dmlStream.Start()
```
Comment (bigsheeper) [inline internal/querynode/flow_graph_query_node.go:214]: Dml stream has been started in `internal/util/flowgraph/input_node.go:37`.

File: internal/querynode/query_collection.go [hunk 4]
```diff
@@ -141,34 +142,54 @@ func (q *queryCollection) close() {
 	q.globalSegmentManager.close()
 }
 
+func (q *queryCollection) getCollectionInReplica(collectionID int64) (*Collection, error) {
```
Comment (bigsheeper) [inline internal/querynode/query_collection.go:145]: I think we should check if collection exists in both streaming and historical.

File: internal/querynode/query_collection.go [hunk 5]
```diff
@@ -141,34 +142,54 @@ func (q *queryCollection) close() {
...
+			if err != nil {
+				return err
+			}
 		}
+		log.Debug("register tSafe watcher and init watcher select case",
+			zap.Any("collectionID", historicalCollection.ID()),
+			zap.Any("delta channels", historicalCollection.getVChannels()))
+	}
+	if err1 != nil && err2 != nil {
+		return fmt.Errorf("collection cannot be found in both historical and streaming replica")
```
Comment (bigsheeper) [inline internal/querynode/query_collection.go:190]: I think we should check if collection is exists in both streaming and historical.
Collection would be created in both streaming and historical even though there is dm data in streaming and no dm data in historical.

File: internal/querynode/query_collection.go [hunk 6]
```diff
@@ -942,22 +963,26 @@ func (q *queryCollection) search(msg queryMsg) error {
 	hisSearchResults, sealedSegmentSearched, err1 := q.historical.search(searchRequests, collection.id, searchMsg.PartitionIDs, plan, travelTimestamp)
 	if err1 != nil {
 		log.Warn(err1.Error())
-		return err1
```
Comment (bigsheeper) [inline internal/querynode/query_collection.go:945]: Why not return error？Even if there is no dm data in historical, search shouldn't return error.

File: internal/querynode/query_collection.go [hunk 7]
```diff
@@ -1167,7 +1192,6 @@ func (q *queryCollection) retrieve(msg queryMsg) error {
 	hisRetrieveResults, sealedSegmentRetrieved, err1 := q.historical.retrieve(collectionID, retrieveMsg.PartitionIDs, q.vectorChunkManager, plan)
 	if err1 != nil {
 		log.Warn(err1.Error())
-		return err1
```
Comment (bigsheeper) [inline internal/querynode/query_collection.go:1170]: same as above

File: internal/querynode/task.go [hunk 8]
```diff
@@ -146,12 +152,6 @@ func (w *watchDmChannelsTask) Execute(ctx context.Context) error {
 			return err
 		}
 	}
-	if hasCollectionInHistorical := w.node.historical.replica.hasCollection(collectionID); !hasCollectionInHistorical {
```
Comment (bigsheeper) [inline internal/querynode/task.go:149]: I don't get it.

File: internal/querynode/task.go [hunk 9]
```diff
@@ -430,12 +568,45 @@ func (r *releaseCollectionTask) PreExecute(ctx context.Context) error {
...
+	if hasCollectionInHistorical {
+		err := r.releaseReplica(r.node.historical.replica, r.node.dataSyncService.removeCollectionDeltaFlowGraph)
+		if err != nil {
+			err2 = errors.New(errMsg + err.Error())
+		} else {
+			// release global segment info
+			r.node.historical.removeGlobalSegmentIDsByCollectionID(r.req.CollectionID)
+		}
+	}
+	if !hasCollectionInStreaming && !hasCollectionInHistorical {
```
Comment (bigsheeper) [inline internal/querynode/task.go:591]: should return nil

File: internal/querycoord/task_scheduler.go [hunk 10]
```diff
@@ -662,7 +662,7 @@ func (scheduler *TaskScheduler) scheduleLoop() {
 			childTasks := triggerTask.getChildTask()
 			if len(childTasks) != 0 {
 				// process loadSegment before watchDmChannel, avoid delete not taking effect
-				highPriorityTasks, lowPriorityTasks := sortInternalTaskByPriority(childTasks, commonpb.MsgType_LoadSegments)
+				highPriorityTasks, lowPriorityTasks := sortInternalTaskByPriority(childTasks, commonpb.MsgType_LoadSegments, commonpb.MsgType_WatchDmChannels)
```
Comment (xige-16) [inline internal/querycoord/task_scheduler.go:665]: There should be no dependency between watchQuerychannel and watchDelta/dm channel

File: internal/querynode/query_collection.go (hunk 5 above)
Comment (xige-16) [reply internal/querynode/query_collection.go:190]: > I think we should check if collection is exists in both streaming and historical. Collection would be created in both streaming and historical even though there is dm data in streaming and no dm data in historical.

It is not necessary to check whether there is this collection in historical or streaming. You can call queryCollection.addTSafeWatcher to add vchannel when you watchDmchane/deltaChannel after watchQueryChannel.

File: internal/querynode/collection.go [hunk 11]
```diff
@@ -135,6 +138,57 @@ func (c *Collection) getPChannels() []Channel {
 	return c.pChannels
 }
 
+// addPChannels add physical channels to physical channels of collection
+func (c *Collection) addPDeltaChannels(channels []Channel) {
+OUTER:
+	for _, dstChan := range channels {
+		for _, srcChan := range c.pDeltaChannels {
```
Comment (cydrain) [inline internal/querynode/collection.go:145]: can use map[][] here

**Pull request merged**